JAVA-764: LWT / CAS Consistency bug on Retry Policy.

adutra opened the pull request:
This is based on top of #587.

Amend all retry policies and handle serial consistency levels accordingly (rethrow should be the preferred decision).

Commit:
JAVA-764: Retry with the normal consistency level (not the serial one) when a write times out on the Paxos phase.

tolbertam commented on driver-core/src/test/java/com/datastax/driver/core/policies/CASRetryPolicyIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+        session.execute("CREATE TABLE foo(k int primary key, c int)");
+        session.execute("INSERT INTO foo(k, c) VALUES (0, 0)");
+
+        Statement s = new SimpleStatement("UPDATE foo SET c = 1 WHERE k = 0 IF EXISTS")
```
This will fail with C\* 1.2 because it doesn't support CAS operations, can you add `@CassandraVersion(major=2.0)` to this?

adutra replied:
Thanks, good catch! I also modified the statement because UPDATE IF EXISTS is only possible since CASSANDRA-8610.

olim7t commented:
We're losing some information though, namely the fact that it's the paxos phase, not the mutation phase, that timed out. If we do the switch in the query handler, the retry policy will have no way to differentiate the two cases.
And the default behavior if the paxos phase timed out should probably be to not retry. If a quorum of nodes was not reachable to establish a proposal, it's unlikely that switching to the next host would fix the issue.

Commit:
JAVA-982: Introduce new method ConsistencyLevel.isSerial().

adutra commented:
Rebased on top of current JAVA-1050. Addressed @olim7t concerns.

Summary of recent changes:
- New method `ConsistencyLevel.isSerial()`;
- Amended `onUnavailable` implementations of `DefaultRetryPolicy`and `DowngradingConsistencyRetryPolicy` to rethrow on serial consistency levels (see a12838c);
- It turns out that `onWriteTimeout` implementations need not be amended (or so I mistakenly assumed);
- Javadoc improvements to clarify the meaning of consistency levels.

tolbertam commented:
> New method ConsistencyLevel.isSerial();

Nice, that also takes care of [JAVA-982](https://datastax-oss.atlassian.net/browse/JAVA-982)

tolbertam commented:
Now that #587 is merged, can we rebase this on 2.1?

olim7t commented:
I've rebased on 2.1 and removed the commits related to our initial approach. This will require a bit of cleanup/squashing before merging.

olim7t commented:
After giving it more thought, I think there is no reason not to retry on the next host when the CL is serial: the "network-isolated coordinator" hypothesis applies as well.
I pushed a commit to try the next host with the statement's original CL.

olim7t commented:
One more change: retry with the previous CL instead of the original CL. We don't have the case in default policies, but the CL might already have been downgraded in a previous retry.

adutra commented:
lgtm :+1: 
Small note before we squash commits together: let's try to keep the commit for JAVA-982 separate.

olim7t commented:
Done.

tolbertam commented on driver-core/src/test/java/com/datastax/driver/core/policies/CASRetryPolicyIntegrationTest.java:
```diff
@@ -0,0 +1,103 @@
+ * Tests for retry policies in combination with lightweight transactions.
+ * <p/>
+ * Note: we can't extend {@link AbstractRetryPolicyIntegrationTest} here, because we are
+ * testing LWT statements and Scassandra doesn't support priming consistency levels yet.
```
I'll add support for that to scassandra and create a branch with tests.

tolbertam replied:
I think these tests need to be changed now that we retry once on unavailable, and then there are no remaining hosts to try, so we get an NHAE

tolbertam commented:
Opened scassandra/scassandra-server#143 to add the capability to simulate errors returning CL other than what the query specified to help test this.  Created [java764tests](https://github.com/datastax/java-driver/tree/java764tests) branch with added tests which pass against this branch that will open PR for when version of scassandra released with this feature.
